include/AmplitudeBasis.h

```diff
@@ -23,133 +23,331 @@
+    {}
+
+    /// constructor
+    /// \param c vector of amplitudes
+    /// \param cov 3x3 covariance matrix of 2x2 covariances (between real and imaginary parts of amplitudes), defaults to 0
+    explicit canonical(Vector<std::complex<T>, 3> c,
+            covariance_matrix<T> cov = covariance_matrix<T>()) :
+        basis<T>(c, cov)
+    {}
+
+    /// constructor
+    /// \param cov 3x3 covariance matrix of 2x2 covariances (between real and imaginary parts of amplitudes)
+    explicit canonical(const covariance_matrix<T>& cov) :
+        basis<T>(Vector<std::complex<T>, 3>({0, 0, 0}), cov)
+    {}
+
+    /// constructor
+    /// \param c1 S amplitude with covariance
+    /// \param c2 P amplitude with covariance
+    /// \param c3 D amplitude with covariance
+    canonical(const complex_basis::cartesian<T>& c1, const complex_basis::cartesian<T>& c2, const complex_basis::cartesian<T>& c3) :
+        basis<T>(Vector<std::complex<T>, 3>({c1, c2, c3}),
+                 diagonalMatrix<complex_basis::covariance_matrix<T>, 3>({c1.covariance(), c2.covariance(), c3.covariance()}))
+    {}
 
     /// casting constructor
     explicit constexpr canonical(const transversity<T>& t) :
-        S(sqrt(2. / 3.) * t.parallel - sqrt(1. / 3.) * t.longitudinal),
-        P(t.perpendicular),
-        D(sqrt(1. / 3.) * t.parallel + sqrt(2. / 3.) * t.longitudinal)
+        basis<T>(t, jacobian_t_to_c)
     {}
 
     /// casting constructor
     explicit constexpr canonical(const helicity<T>& h) :
-        S(sqrt(1. / 3.) * (h.plus + h.minus - h.zero)),
-        P(sqrt(1. / 2.) * (h.plus - h.minus)),
-        D(sqrt(1. / 6.) * (h.plus + h.minus) + sqrt(2. / 3.) * h.zero)
+        basis<T>(h, jacobian_h_to_c)
     {}
 
+    /// \return S amplitude (l=0)
+    const std::complex<T> s() const
+    { return basis<T>::coordinates()[0]; }
+
+    /// \return P amplitude (l=1)
+    const std::complex<T> p() const
+    { return basis<T>::coordinates()[1]; }
+
+    /// \return D amplitude (l=2)
+    const std::complex<T> d() const
+    { return basis<T>::coordinates()[2]; }
+
     /// access canonical amplitude via angular momentum
-    /// \param L angular momentum to retrieve amplitude for
-    const std::complex<T>& operator[] (size_t L)
+    /// \param l angular momentum to retrieve amplitude for
+    const std::complex<T> operator[] (size_t l) const
     {
-        switch (L) {
+        switch (l) {
             case 0:
-                return S;
+                return s();
             case 1:
-                return P;
+                return p();
             case 2:
-                return D;
+                return d();
             default :
-                throw exceptions::Exception("L must be 0, 1, or 2", "canonical::operator[]");
+                throw exceptions::Exception("l must be 0, 1, or 2", "canonical::operator[]");
         }
     }
+
+private:
+    // transformation matrix (jacobian) from transversity to canonical
```

Comment:
some sort of namespacing. Either

```cpp
amplitude_basis::canonical::jacobian::from_transversity;
```

or 
```cpp
amplitude_basis::jacobian::canonical_to_transversity;
```

My vote is for the second option, since it would match the forms used in `complex_basis`